src/AppInstallerRepositoryCore/RepositorySource.cpp

```diff
@@ -34,52 +34,127 @@ namespace AppInstaller::Repository
             return ISourceFactory::GetForType(details.Type)->Create(details);
         }
 
+        std::chrono::milliseconds GetMillisecondsToWait(std::chrono::seconds retryAfter, size_t randomMultiplier = 1)
+        {
+            if (retryAfter != 0s)
+            {
+                return std::chrono::duration_cast<std::chrono::milliseconds>(retryAfter);
+            }
+            else
+            {
+                // Add a bit of randomness to the retry wait time
+                std::default_random_engine randomEngine(std::random_device{}());
+                std::uniform_int_distribution<long long> distribution(2000, 10000);
+
+                return std::chrono::milliseconds(distribution(randomEngine) * randomMultiplier);
+            }
+        }
+
+        bool IsUpdateSuppressed(const SourceDetails& details)
+        {
+            return std::chrono::system_clock::now() < details.DoNotUpdateBefore;
+        }
+
+        struct AddOrUpdateResult
+        {
+            bool UpdateChecked = false;
+            bool MetadataWritten = false;
+        };
+
         template <typename MemberFunc>
-        bool AddOrUpdateFromDetails(SourceDetails& details, MemberFunc member, IProgressCallback& progress)
+        AddOrUpdateResult AddOrUpdateFromDetails(SourceDetails& details, MemberFunc member, IProgressCallback& progress)
         {
-            bool result = false;
+            AddOrUpdateResult result;
+
+            // Do not update if we are still before the update block time.
+            if (IsUpdateSuppressed(details))

```

Comment:
I had intentionally left it open for a manual update to still attempt this.  We are caught in the middle of respecting the user's request and the server's.  I wouldn't mind it being "harder" for the client to do (aka `--force`), but I feel like we should leave that option open.

Reply:
This is still here...